src/JSInterop/Microsoft.JSInterop/src/JSRuntimeBase.cs

```diff
@@ -88,33 +83,32 @@ internal void EndInvokeDotNet(string callId, bool success, object resultOrExcept
 
             // We pass 0 as the async handle because we don't want the JS-side code to
             // send back any notification (we're just providing a result for an existing async call)
-            BeginInvokeJS(0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", Json.Serialize(new[] {
-                callId,
-                success,
-                resultOrException
-            }, ArgSerializerStrategy));
+            var args = JsonSerializer.ToString(new[] { callId, success, resultOrException }, JsonSerializerOptions);
+            BeginInvokeJS(0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", args);
         }
 
-        internal void EndInvokeJS(long asyncHandle, bool succeeded, object resultOrException)
+        internal void EndInvokeJS(long asyncHandle, bool succeeded, JSAsyncCallResult asyncCallResult)
         {
-            if (!_pendingTasks.TryRemove(asyncHandle, out var tcs))
+            using (asyncCallResult?.JsonDocument)
             {
-                throw new ArgumentException($"There is no pending task with handle '{asyncHandle}'.");
-            }
-
-            if (succeeded)
-            {
-                var resultType = TaskGenericsUtil.GetTaskCompletionSourceResultType(tcs);
-                if (resultOrException is SimpleJson.JsonObject || resultOrException is SimpleJson.JsonArray)
+                if (!_pendingTasks.TryRemove(asyncHandle, out var tcs))
                 {
-                    resultOrException = ArgSerializerStrategy.DeserializeObject(resultOrException, resultType);
+                    throw new ArgumentException($"There is no pending task with handle '{asyncHandle}'.");
                 }
 
-                TaskGenericsUtil.SetTaskCompletionSourceResult(tcs, resultOrException);
-            }
-            else
-            {
-                TaskGenericsUtil.SetTaskCompletionSourceException(tcs, new JSException(resultOrException.ToString()));
+                if (succeeded)
+                {
+                    var resultType = TaskGenericsUtil.GetTaskCompletionSourceResultType(tcs);
+                    var result = asyncCallResult != null ?
+                        JsonSerializer.Parse(asyncCallResult.JsonElement.GetRawText(), resultType, JsonSerializerOptions) :
```

Comment:
There's a feature request to get the serializer to support `JsonElement` which I plan on tackling in my free time. It would avoid all these round-tripping of JsonElement (utf-8 bytes) -> string (utf-16 chars) -> Object and should be a decent perf improvement

Reply:
Cool